### ISSUE TITLE
Cropping is not pixel accurate (JavaImageServlet + JavaImageEditor)

### ISSUE THREAD

**Issue by Typografikon** [perfectsense/brightspot-cms#323]
How to reproduce:

1. Create 2800x1867px image  and upload it to cms
2. Request cropped thumbnail with **650x350px** dimensions, use fillArea option
3. You got image with **650x351px** dimensions, which is problem in pixel-accurate layouts

Possible workarounds/solutions:

1. add 5% to resize so subsequent crop always crop small place around image (see solution1.txt patch)
2. add one more resize with ignoreAspectRatio option after image is cropped (see solution2.txt patch)

Solution 1 perform 5% image zoom so solution 2 is probably more appropriate but requires additional resize. Both solutions are more likely workarounds, not really solutions. Proper solution should improve dimensions calulations or use third-party library for it. I usually use https://github.com/coobird/thumbnailator for this tasks because there is no need to calculate dimensions for various opearations.

**Comment by Typografikon** [comment]
Cannot attach txt files so pasting patches here : 

1. **solution1.txt:**

```
--- old/JavaImageServlet.java
+++ new/JavaImageServlet.java
@@ -288,8 +288,8 @@
                     int resizeWidth = width;
 
                     if (option == null || !option.equals(ImageEditor.RESIZE_OPTION_IGNORE_ASPECT_RATIO)) {
-                        resizeHeight = (int) ((double) bufferedImage.getHeight() / (double) bufferedImage.getWidth() * (double) width);
-                        resizeWidth  = (int) ((double) bufferedImage.getWidth() / (double) bufferedImage.getHeight() * (double) height);
+                        resizeHeight = (int) (((double) bufferedImage.getHeight() / (double) bufferedImage.getWidth()  + 0.05) * (double) width);
+                        resizeWidth  = (int) (((double) bufferedImage.getWidth()  / (double) bufferedImage.getHeight() + 0.05) * (double) height);
                     }
 
                     bufferedImage = javaImageEditor.reSize(bufferedImage, resizeWidth, resizeHeight, option, quality);
```

2. **solution2.txt**

```
--- old/JavaImageServlet.java
+++ new/JavaImageServlet.java
@@ -317,6 +317,7 @@
 
                         if (width <= bufferedImage.getWidth() && height <= bufferedImage.getHeight()) {
                             bufferedImage = javaImageEditor.crop(bufferedImage, x, y, width, height);
+                            bufferedImage = javaImageEditor.reSize(bufferedImage, width, height, ImageEditor.RESIZE_OPTION_IGNORE_ASPECT_RATIO, quality);
                         }
                     }
```

**Comment by gitastrophe** [comment]
Typografikon, I am unable to reproduce this issue on the current 3.2-SNAPSHOT versions of Dari and Brightspot.

My settings for the StandardImageSize used are:

![screen shot 2015-10-02 at 2 45 12 pm](https://cloud.githubusercontent.com/assets/400882/10254833/3def6bee-6914-11e5-8647-0f7478ba7fe8.png)

I uploaded a 2800x1867 file as my test StorageItem.  The output of ImageTag.Builder#toUrl produces a URL for an image file with dimension exactly 650x350:

http://brendan.local:8080/_image/thumbnail/650x350^/resize/650x350^/?url=%2Flocal-storage%2Ffd%2F4b%2Fecf7682c4406acc8133780218db7%2Fresizes%2F1500%2Fautumn-trees-in-dresden-scaled.jpeg

Can you please provide more information for me to reproduce the issue?

Which versions of Brightspot and Dari are you using?
What are the exact settings in your StandardImageSize?
Are you using any context.xml settings to alter the behavior of the JavaImageEditor?

**Comment by Typografikon** [comment]
I made some tests and I cannot reproduce it with empty database and 3.0- or 3.1-SNAPSHOT (using archetype). But I can 100% reproduce it in my project. I use 3.0-SNAPSHOT I have no idea what is wrong, I use no special configuration for JavaImageEditor, image  rendering with this jsp (item.image.file is StorageItem):

```jsp
<cms:img
      src="${item.image.file}"
      alt="${item.image.title}"
      size="topicsImage"
      hideDimensions="true"
      class="img-responsive"  />
```
Image url is

```
http://as3.brightspotcms.localhost:8080/_image/thumbnail/650x350%5E/resize/650x350%5E/?url=http://localhost:8080/storage-mysql/57/14/914e70874c87b016653e73f6e8fd/waldorfska-pedagogika.jpg
```
and image setting:

![settings_cms_brightspot_2015-11-06_22-21-24](https://cloud.githubusercontent.com/assets/13291261/11009259/f4fb7814-84d5-11e5-84ef-417985df7758.png)

Looks like I must investigate it a little more...

**Comment by Typografikon** [comment]
I got it - reproduced with brightspot 3.1:

First of all follow _Quick Installation steps_ from http://www.brightspot.com/docs/3.1/overview/installation#quick-installation:

```
mvn archetype:generate -B \
    -DarchetypeRepository=http://public.psddev.com/maven \
    -DarchetypeGroupId=com.psddev \
    -DarchetypeArtifactId=cms-app-archetype \
    -DarchetypeVersion=3.1.4461-c7651a \
    -DgroupId=com.project \
    -DartifactId=projectName
```

Then add these files to project:

projectName/src/main/java/com/project/Picture.java:

```java
package com.project;

import com.psddev.cms.db.Content;
import com.psddev.cms.db.ToolUi.Filterable;
import com.psddev.cms.db.ToolUi.Note;
import com.psddev.cms.db.ToolUi.Placeholder;
import com.psddev.cms.db.ToolUi.Sortable;
import com.psddev.cms.db.ToolUi.Tab;
import com.psddev.dari.db.Recordable.DisplayName;
import com.psddev.dari.db.Recordable.PreviewField;
import com.psddev.dari.util.StorageItem;
import com.psddev.cms.db.Renderer.*;

@PreviewField("file")
@DisplayName("Picture")
@Note("Picture...")

@Path("/WEB-INF/picture.jsp")
@LayoutPath("/WEB-INF/layout.jsp")
public class Picture extends Content {

	@Indexed
	@Required
	@Sortable
	@Tab("Picture")
	@DisplayName("Picture title")
	@Placeholder("Picture title...")
	protected String title;
	
	@Indexed
	@Required
	@Filterable(false)
	@Tab("Picture")
	@DisplayName("Picture file")
	protected StorageItem file;

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public StorageItem getFile() {
		return file;
	}

	public void setFile(StorageItem file) {
		this.file = file;
	}
	
}
```
projectName/src/main/webapp/WEB-INF/picture.jsp:
```jsp
<%@ page contentType="text/html; charset=UTF-8" %>
<%@ taglib prefix="c"   uri="http://java.sun.com/jsp/jstl/core" %>
<%@ taglib prefix="fn"  uri="http://java.sun.com/jsp/jstl/functions" %>
<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
<%@ taglib prefix="cms" uri="http://psddev.com/cms" %>

<h1><cms:render value="${content.title}" /></h1>
<cms:img src="${content.file}" alt="${content.title}" title="${content.title}" size="tile" />
```

projectName/src/main/webapp/WEB-INF/layout.jsp:

```jsp
<%@ page contentType="text/html; charset=UTF-8" %>
<%@ taglib prefix="c"   uri="http://java.sun.com/jsp/jstl/core" %>
<%@ taglib prefix="fn"  uri="http://java.sun.com/jsp/jstl/functions" %>
<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
<%@ taglib prefix="cms" uri="http://psddev.com/cms" %>

<c:if test="${not empty content}">
	<cms:render value="${content}" />
</c:if>
```

Now fire up project 

```
mvn -P run clean package cargo:run
```

and set _tile_ image size as 650x350 with fill area option:

![settings_cms_brightspot_2015-11-06_23-50-32](https://cloud.githubusercontent.com/assets/13291261/11010705/3281a95a-84e1-11e5-8b2c-926cf1452b5c.png)

In cms http://localhost:9480/cms/ create new image instance uploading jpeg image with dimension *1081x650* (notice 650 is the same size as in image _tile_ setting) and assign also permalink URLs (I used /) :

![picture_cms_brightspot_2015-11-07_00-10-43](https://cloud.githubusercontent.com/assets/13291261/11010997/04d224e6-84e4-11e5-95c9-9dd0789a3caf.png)

Finally load page http://localhost:9480/ and inspect native image dimensions:

![snimek obrazovky](https://cloud.githubusercontent.com/assets/13291261/11011020/55a18a60-84e4-11e5-9b7e-b9ce1e0d15ff.png)

You see 650px x 351px instead of 650px x 350px. 

So is it :-) Can you reproduce it now?

**Comment by rhino88** [comment]
@Typografikon can you send us the raw json data of your image object?

You can find this easily from the CMS by going to the developer tools on the Edit page in the CMS, and clicking on "View Raw Data".

![image](https://cloud.githubusercontent.com/assets/1299507/11345921/fe4d7ccc-91e7-11e5-938a-b55f7eb37f70.png) 

We believe this is not an issue in ImageTag, but it actually may be a rounding issue from the CMS Image Editor.

Thanks!

**Comment by Typografikon** [comment]
Sure it is JavaImageServlet or JavaImageEditor issue. 

Here is raw json data of my testing image:

```json
{
  "title" : "Test",
  "file" : {
    "storage" : "projectName.local",
    "path" : "e3/01/25ae17af495c960f49b228116de6/test.jpg",
    "contentType" : "image/jpeg",
    "metadata" : {
      "cms.edits" : {
      },
      "originalFilename" : "test3-1081x650.jpg",
      "http.headers" : {
        "Cache-Control" : [ "public, max-age=31536000" ],
        "Content-Length" : [ "274417" ],
        "Content-Type" : [ "image/jpeg" ]
      },
      "width" : 1081,
      "height" : 650,
      "cms.crops" : {
      },
      "cms.focus" : {
      },
      "resizes" : [ {
        "storage" : "projectName.local",
        "path" : "e3/01/25ae17af495c960f49b228116de6/resizes/500/test.jpg",
        "contentType" : "image/jpeg",
        "metadata" : {
          "width" : 500,
          "height" : 301
        }
      } ]
    }
  },
  "cms.content.publishDate" : 1446849464172,
  "cms.content.publishUser" : {
    "_ref" : "00000150-def1-d858-abfc-fff78b9d0000",
    "_type" : "00000150-deed-d858-abfc-ffff5cd9003a"
  },
  "cms.content.updateDate" : 1446851395691,
  "cms.content.updateUser" : {
    "_ref" : "00000150-def1-d858-abfc-fff78b9d0000",
    "_type" : "00000150-deed-d858-abfc-ffff5cd9003a"
  },
  "cms.ab.fields" : {
  },
  "cms.seo.keywords" : [ ],
  "cms.seo.robots" : [ ],
  "cms.directory.automaticRawPaths" : [ ],
  "cms.directory.paths" : [ "00000150-def5-dea0-abfe-ffff7cc60000/index" ],
  "cms.directory.pathTypes" : {
    "00000150-def5-dea0-abfe-ffff7cc60000/index" : "PERMALINK"
  },
  "cms.directory.pathsMode" : "MANUAL",
  "_id" : "00000150-def2-d858-abfc-fff7833c0000",
  "_type" : "00000150-def0-d858-abfc-fff7c62a0000"
}
```

If you need any other informations feel free to ask me.

**Comment by Typografikon** [comment]
Fixed in [v3.2.5001-98eff8](https://github.com/perfectsense/brightspot-cms/releases/tag/v3.2.5001-98eff8) Thanks!
